weasyl/recommendation.py

```diff
@@ -0,0 +1,243 @@
+
+
+def _fast_similarity(ratings, kind='user'):
+    """
+    Calculate user to user or item to item similarities.
+    This function can take a few seconds to run.
+    """
+    # The function below was adapted from the blog post linked above. See:
+    # http://blog.ethanrosenthal.com/2015/11/02/intro-to-collaborative-filtering/#Collaborative-filtering
+    #
+    # With some minor alterations:
+    #   - We don't use epsilon because scipy dies when we try to add a scalar to a sparse matrix.
+    #     This shouldn't be matter because all users in the matrix have at least one favorite.
+    #   - I use sim.diagonal() instead of np.diag() because sparse matrices complain mightily when
+    #     it comes to np.diag().
+    #   - Things die if we try to do regular division. Instead, we construct a diagonal matrix with
+    #     the reciprocals of everything we would have divided with and both pre-multiply (to scale
+    #     every row by the first user) and post multiply (to scale every column by the second user).
+    if kind == 'user':
+        sim = ratings.dot(ratings.T)
+    elif kind == 'item':
+        sim = ratings.T.dot(ratings)
+    norms = np.array([np.sqrt(sim.diagonal())])
+    norms_sparse_diag = sparse.diags(1/norms.ravel(), format='csr')
+    return (norms_sparse_diag * sim * norms_sparse_diag)
+
+
+def select_list(userid, rating):
+    """
+    Get the top recommendations for a user and select their submission info. Doesn't show hidden
+    content or content that violates friends rules.
+    """
+    submitids = recs_for_user(userid)
+    if not submitids:
+        return submitids
+    # TODO(hyena): This is grossly adapted from submission.py. Commit less SQL violence.
+    statement = [
+        "SELECT su.submitid, su.title, su.rating, su.unixtime, "
+        "su.userid, pr.username, su.settings, su.subtype "]
+    statement.extend(submission.select_query(userid=userid, rating=rating))
+    statement.append(" AND su.submitid IN %(recs)s")
+
+    items = {i[0]: {
+                 "contype": 10,
+                 "submitid": i[0],
+                 "title": i[1],
+                 "rating": i[2],
+                 "unixtime": i[3],
+                 "userid": i[4],
+                 "username": i[5],
+                 "subtype": i[7],
+             } for i in d.engine.execute("".join(statement), recs=tuple(submitids))}
+    query = [items[x] for x in submitids if x in items]  # Re-sort.
+    media.populate_with_submission_media(query)
+
+    return query
+
+
+@d.record_timing
+def recs_for_user(userid, k=20, count=100):
+    """
+    Generates recommendations for a weasyl user.
+    Will not include the user's own submissions or items they've rated.
+
+    Args:
+        userid (int): The weasyl userid.
+        k (int, optional): How many closest other users to use. Defaults to 20.
+        count (int, optional): How many items to return for the user. Defaults to 10.
```

Comment:
``Defaults to 10.`` => ``Defaults to 100.`` -- The function definition currently has ``count`` default to 100.